stratum/lib/security/certificate.cc

```diff
@@ -91,24 +93,79 @@ util::Status GenerateUnsignedCert(X509* unsigned_cert,
       name, "CN", MBSTRING_UTF8,
       reinterpret_cast<const unsigned char*>(common_name.c_str()), -1, -1, 0));
 
-  return util::OkStatus();
+  RETURN_OK();
+}
+
+util::Status AddExtension(X509* cert, X509V3_CTX* ctx, const std::string& name,
+                          const std::string& value) {
+  bssl::UniquePtr<X509_EXTENSION> ext(
+      X509V3_EXT_nconf(nullptr, ctx, const_cast<char*>(name.c_str()),
+                       const_cast<char*>(value.c_str())));
+
+  // Return error if the extension could not be created
+  if (!ext.get()) {
+    bssl::UniquePtr<BIO> bio(BIO_new(BIO_s_mem()));
+    CHECK_RETURN_IF_FALSE(bio.get()) << "Failed to allocate string buffer.";
+    ERR_print_errors(bio.get());
+
+    BUF_MEM* mem = nullptr;
+    BIO_get_mem_ptr(bio.get(), &mem);
+    if (mem != nullptr && mem->data && mem->length) {
+      RETURN_ERROR(ERR_INVALID_PARAM).without_logging()
+          << "Could not add extension " << name << " with value: " << value
+          << ".\n"
+          << std::string(mem->data, mem->length);
+    }
+  }
+
+  // Add the extension to the end of the list
+  CHECK_RETURN_IF_FALSE(X509_add_ext(cert, ext.get(), -1));
+
+  RETURN_OK();
```

Comment:
First time seeing this. wow

Reply:
I was tired of the seg faults with no errors. 😂

Turns out most OpenSSL commands will write to an error buffer before returning; we just need to read them. We could use in other places that we make OpenSSL calls, but here's one way to get some of the error data.

Reply:
And if the wow was for how complicated it is to read a string from OpenSSL... I agree.

We can (should) probably add a helper that takes a BIO and returns a string.